doc/sphinx/proof-engine/ltac.rst

```diff
@@ -1321,16 +1314,53 @@ Pattern matching on goals and hypotheses: match goal
       | [ {*, @match_hyp } |- @match_pattern ]
       | _
       match_hyp ::= @name : @match_pattern
-      | @name := {? [ @match_pattern ] : } @match_pattern
+      | @name := @match_pattern
+      | @name := [ @match_pattern ] : @match_pattern
 
    :tacn:`lazymatch goal`, :tacn:`match goal` and :tacn:`multimatch goal` are :token:`l1_tactic`\s.
 
    Use this form to match hypotheses and/or goals in the proof context.  These patterns have zero or
    more subpatterns to match hypotheses followed by a subpattern to match the conclusion.  Except for the
-   differences noted below, this works the same as the corresponding `*match* term` construct (see :tacn:`match`).
+   differences noted below, this works the same as the corresponding :n:`@match_key @ltac_expr` construct
+   (see :tacn:`match`).  Each current goal is processed independently.
 
    :n:`{*, @match_hyp }`
-      Patterns to match with hypotheses.  All must match distinct hypotheses in order for the branch to match.
+      Patterns to match with hypotheses.  Each pattern must match a distinct hypothesis in order
+      for the branch to match.
+
+      Hypotheses have the form :n:`@name {? := @term__binder } : @type`.  Patterns bind each of
+      these nonterminals separately:
+
+      .. list-table::
+         :widths: 2 1
+         :header-rows: 1
+
+         * - Pattern syntax
+           - Example pattern
+
+         * - :n:`@name : @match_pattern__type`
+           - `n : ?t`
+
+         * - :n:`@name := @match_pattern__binder`
+           - `n := ?b`
+
+         * - :n:`@name := @term__binder : @type`
+           - `n := ?b : ?t`
+
+         * - :n:`@name := [ @match_pattern__binder ] : @match_pattern__type`
+           - `n := [ ?b ] : ?t`
+
+      ..
+
+         :token:`name` can't have a `?`.  Note that the last two forms are equivalent except that:
+
+         - if the `:` in the third form has been bound to something else in a notation, you must use the fourth form.
+           Note that cmd:`Require Import` `ssreflect` loads a notation that does this.
+         - a :n:`@term__binder` in the form `[ ?l ]` must be parenthesized or, for the fourth form, use double brackets:
```

Comment:
Perhaps
```suggestion
         - a :n:`@term__binder` in the form `[ ?l ]` (e.g., denoting a singleton list after :cmd:`Import` `ListNotations`) must be parenthesized or, for the fourth form, use double brackets:
```
?

Reply:
I'll use your wording for now.  Maybe revisit this after the pattern matching and unification sections have been written.  However:  `[ ?l ]` won't match `[ a; b]`??  That's not what I'd expect.  Is that a bug or a feature?

Reply:
`[ ?l ]` will match everything, including `[a; b]` and `[]` and `1` and `nat`.  This is the surprising thing.  By contrast, if I write `[ ?l ]` in a pattern elsewhere, it only matches the singleton list.

Reply:
I should have said `[ [ ?l ] ]` or `( [ ?l ] )`.  They will match `[a; b]` and `[]` but not `1` and `nat`, right?

Reply:
@JasonGross says: Don't have my computer on my right now, so replying by email (feel free to splice this info the thread).

No, `([?l])` is syntactic sugar for the pattern `cons ?l nil`, which does not match `[]` (`nil`) nor `[x; y]` (`cons x (cons y nil)`).